src/server/pfs/server/driver.go

```diff
@@ -1461,35 +1454,83 @@ func (d *driver) createBranch(ctx context.Context, branch *pfs.Branch, commit *p
 				// possible that branch exists but has no head commit. This is fine, but
 				// branchInfo.Head must also be nil
 				if _, ok := err.(pfsserver.ErrNoHead); !ok {
-					return err
+					return fmt.Errorf("inspectCommit: %v", err)
 				}
+				commit = nil
+			} else {
+				commit = commitInfo.Commit
 			}
 		}
 		branchInfo := &pfs.BranchInfo{
 			Branch: branch,
 			Name:   branch.Name,
 		}
-		if commitInfo != nil {
-			branchInfo.Head = commitInfo.Commit
+		// Get the previous version of this branch if there is one, we do this
+		// so that we don't lose pre existing subvenance.
+		if err := d.branches(branch.Repo.Name).ReadWrite(stm).Get(branch.Name, branchInfo); err != nil && !col.IsErrNotFound(err) {
+			return err
 		}
-
-		if len(provMap) > 0 {
+		branchInfo.Head = commit
+		branchInfo.DirectProvenance = nil
+		for _, provBranch := range provenance {
+			(*branchSet)(&branchInfo.DirectProvenance).add(provBranch)
+		}
+		// We don't just need to update (or create) branch's Provenance but
+		// also the Provenance of all our Subvenance (in the case of an update)
+		toUpdate := []*pfs.BranchInfo{branchInfo}
+		for _, subvBranch := range branchInfo.Subvenance {
+			subvBranchInfo := &pfs.BranchInfo{}
+			if err := d.branches(subvBranch.Repo.Name).ReadWrite(stm).Get(subvBranch.Name, subvBranchInfo); err != nil {
+				return err
+			}
+			toUpdate = append(toUpdate, subvBranchInfo)
+		}
+		// Sorting is important here because it sorts topologically. This means
+		// that when evaluating element i of `toUpdate` all elements < i will
+		// have already been evaluated and thus we can safely use their
+		// Provenance field.
+		sort.Slice(toUpdate, func(i, j int) bool { return len(toUpdate[i].Provenance) < len(toUpdate[j].Provenance) })
+		for _, branchInfo := range toUpdate {
+			oldProvenance := branchInfo.Provenance
+			branchInfo.Provenance = nil
+			for _, provBranch := range branchInfo.DirectProvenance {
+				if err := d.addBranchProvenance(branchInfo, provBranch, stm); err != nil {
+					return err
+				}
+				provBranchInfo := &pfs.BranchInfo{}
+				if err := d.branches(provBranch.Repo.Name).ReadWrite(stm).Get(provBranch.Name, provBranchInfo); err != nil {
+					return err
+				}
+				for _, provBranch := range provBranchInfo.Provenance {
+					if err := d.addBranchProvenance(branchInfo, provBranch, stm); err != nil {
+						return err
+					}
+				}
+			}
+			if err := d.branches(branchInfo.Branch.Repo.Name).ReadWrite(stm).Put(branchInfo.Branch.Name, branchInfo); err != nil {
+				return err
+			}
+			for _, oldProvBranch := range oldProvenance {
+				if !(*branchSet)(&branchInfo.Provenance).has(oldProvBranch) {
+					// Provenance was deleted, so we delete ourselves from their subvenance
```

Comment:
I'm still thinking through this code; is it possible that for a given `oldProvBranch`, there's still a path from `oldProvBranch` to `toUpdate[i].Branch`, but it just doesn't go through the edge `oldProvBranch -> [branch from the argument]`?

If so, it might make sense to add a test for this. I actually think we might need to write this part by putting all the branches in `toUpdate` into two data structures: a stack and a queue. You pull all the branches out of the queue to update their provenance, and then you pull all the branches out of the stack to update their subvenance. We can talk about this in person too, because I'm not 100% sure yet.